src/uct/ib/base/ib_md.c

```diff
@@ -645,26 +672,33 @@ ucs_status_t uct_ib_verbs_mem_reg(uct_md_h uct_md, void *address, size_t length,
                                   const uct_md_mem_reg_params_t *params,
                                   uct_mem_h *memh_p)
 {
-    uct_ib_md_t *md = ucs_derived_of(uct_md, uct_ib_md_t);
+    uct_ib_md_t *md    = ucs_derived_of(uct_md, uct_ib_md_t);
+    uint64_t uct_flags = UCT_MD_MEM_REG_FIELD_VALUE(params, flags,
+                                                    FIELD_FLAGS, 0);
+    int attempt_count  = 0;
     struct ibv_mr *mr_default;
     uct_ib_verbs_mem_t *memh;
     uct_ib_mem_t *ib_memh;
     uint64_t access_flags;
     ucs_status_t status;
 
-    status = uct_ib_memh_alloc(md, length,
-                               UCT_MD_MEM_REG_FIELD_VALUE(params, flags,
-                                                          FIELD_FLAGS, 0),
+    status = uct_ib_memh_alloc(md, length, uct_flags,
                                sizeof(*memh), sizeof(memh->mrs[0]), &ib_memh);
     if (status != UCS_OK) {
         goto err;
     }
 
-    memh         = ucs_derived_of(ib_memh, uct_ib_verbs_mem_t);
-    access_flags = uct_ib_memh_access_flags(&memh->super, md->relaxed_order);
+    memh = ucs_derived_of(ib_memh, uct_ib_verbs_mem_t);
+
+    do {
+        access_flags = uct_ib_memh_access_flags(&memh->super, md->relaxed_order,
+                                                attempt_count > 0, uct_flags);
+
+        status       = uct_ib_reg_mr(md, address, length, params, access_flags,
+                                     NULL, &mr_default, attempt_count == 0);
+        attempt_count++;
```

Comment:
This retry should happen in UCP layer, otherwise IMO it can cause issues with UCP rcache region merges

Reply:
@yosefe Where are the rcache region merges triggered from?
Do you think we should still try to register with full permissions and only then with specific ones?